### ISSUE TITLE
TFJob should work well with pipelines

### ISSUE THREAD

**Issue by jlewi** [kubeflow/pipelines#677]
I'm creating this uber issue to track issues related to making pipelines work well with TFJob. Hopefully this can serve as a model for how to make pipelines work well with other custom resources.

Here's a list of known issues
   #407 TFJob doesn't forward error logs from the jobs
   #408 TFJob doesn't stop trainer jobs after timeout
   #218 provide argument to assign GCP service account

In addition I think there are some additional issues

* The [launcher](https://github.com/kubeflow/pipelines/blob/df635af4099af67e5aae317f2d6794e058eb1279/components/kubeflow/launcher/src/launch_tf_job.py#L126) is getting credentials in a GKE specific way
   * This seems unnecessary; if this code is running in a pod with a service account with suitable RBAC then it should be able to authenticate to the master in a way independent of the Cloud

* The [launcher](https://github.com/kubeflow/pipelines/blob/df635af4099af67e5aae317f2d6794e058eb1279/components/kubeflow/launcher/src/launch_tf_job.py#L184) doesn't seem to be resilient to pod restarts
  
    * The launcher is waiting for the job to finish
    * Jobs could be long running so its possible the pipelines pod will get preempted while waiting for the TFJob to finish
    * It looks like the launcher is using [generateName](https://github.com/kubeflow/pipelines/blob/d9a1313b88d9a0db52792016f8faab91f9cb4bae/components/kubeflow/launcher/src/train.template.yaml#L18) to assign a name to the job.
    * So if the pod gets preempted it will restart and launch another instance of the job with a unique name.

**Comment by jlewi** [comment]
Marking this as P1 for 0.5 because it would be great to have a canonical example of how to integrate with custom resources.

**Comment by amygdala** [comment]
On this point:
> The launcher is getting credentials in a GKE specific way

I don't think this is necessary any more (IIRC, it was used prior to using the tf-job client directly). When I removed it from my own similar code, things worked.

(However, on a related note, adding `.apply(gcp.use_gcp_secret('user-gcp-sa'))` to the launcher step seems to break things, as that apparently does not allow creating cluster resources. Not sure if that's a bug exactly but it seems problematic. https://github.com/kubeflow/pipelines/issues/705 )

**Comment by randxie** [comment]
Adding another issue: when using TPU with tfjob, I found that the tfjob does not reuse allocated cloud TPU (if the pod fails and gets restarted). Instead, it tries to recreate a new TPU instance.

**Comment by jlewi** [comment]
@randxie Can you file a separate issue regarding TFJob and TPU in kubeflow/tf-operator I don't think that has anything to do with TFJob and pipelines.

**Comment by jessiezcc** [comment]
@jlewi looks like most of issues are already closed, what's the remaining work relevant to pipeline that remains open?

**Comment by jlewi** [comment]
@jessiezcc It looks like the pipeline launcher isn't resilient to the pod failing; see my initial comment on this issue. I believe that's still a problem.

**Comment by cyliu0204** [comment]
Hi all:
After reading the comments above , I still cannot figure out how to deploy distribute training with tensorflow or pytorch with pipelines, How is the issue progressing?

**Comment by mak-454** [comment]
Hi, if launcher is to be used for TFjob or for any custom resources. What is the idea to handle volumes? If volumes are used with ContainerOp then the volumes will be attached to launcher and not to the training pods launched by TFJob.

**Comment by mak-454** [comment]
I gather from below discussions,
https://github.com/kubeflow/pipelines/issues/801#issuecomment-462182776
https://github.com/kubeflow/pipelines/issues/1345

that I should use ResourceOp for TFJob with volumes until the PR https://github.com/kubeflow/pipelines/pull/1494 is merged

**Comment by jlewi** [comment]
@mak-454 the current recommendation would be to write some python code to create the TFJob and then execute that code as a step in your graph.

Your python function can attach any volumes as needed to the TFJob.

**Comment by mak-454** [comment]
@jlewi will do as recommended. Thanks.

**Comment by xiaogaozi** [comment]
Looks like mount volume for TFJob require the volume type is [`ReadWriteMany`](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#access-modes) (all pods share same volume), so something like AWS EBS cannot be used in this case (it's `ReadWriteOnce`). It will be helpful if TFJob operator could support mount `ReadWriteOnce` volume.

**Comment by pretidav** [comment]
Dear all, In the end is it possibile to use the tf_job_launcher within a pipeline and retrieve files (i.e. multiple outputs not just a json with scores like katib launcher does) produced by such a job? 
E.g. think about launching a model training within a tf_job I would like to be able to retrieve scores, models and additional outputs from the training to be fed to subsequent components within the pipeline.

**Comment by jlewi** [comment]
@pretidav see my previous comment 
https://github.com/kubeflow/pipelines/issues/677#issuecomment-538617461

It doesn't look like tf_job_launcher.py is being maintained.
https://github.com/kubeflow/pipelines/commits/master/components/kubeflow/launcher

@kubeflow/wg-training-leads What do you want to do about the TFJob Launcher? @hougangliu are you still maintaining the TFJob launcher per the owner file
https://github.com/kubeflow/pipelines/blob/master/components/kubeflow/launcher/OWNERS

**Comment by pretidav** [comment]
Sorry @hougangliu, I noticed that in your e2e example ( https://github.com/kubeflow/pipelines/blob/master/samples/contrib/e2e-mnist/mnist-pipeline.ipynb ) you take advantage of the katib launcher but not the tfjob launcher. 
Would you suggest to do the same or use both launchers? 
(regarding that example I have another issue https://github.com/kubeflow/pipelines/issues/4770 if you want to have a look)

**Comment by Jeffwan** [comment]
I think the plan is to add more native launch support for different training operators. 

This is tracked here but we don't have capacity to make it recently. 

https://github.com/kubeflow/common/blob/master/ROADMAP.md
https://github.com/kubeflow/pipelines/issues/3445

**Comment by jlewi** [comment]
I'm closing this issue. @pretidav If you still have questions please open new more specific issues.
